sys/net/application_layer/asymcute/asymcute.c

```diff
@@ -0,0 +1,937 @@
+                        size_t pos, size_t len)
+{
+    /* verify message length */
+    if (len < (pos + 6)) {
+        return;
+    }
+
+    uint16_t topic_id = byteorder_lebuftohs(&data[pos + 2]);
+
+    /* find any subscription for that topic */
+    mutex_lock(&con->lock);
+    asymcute_sub_t *sub = NULL;
+    for (asymcute_sub_t *cur = con->subscriptions; cur; cur = cur->next) {
+        if (cur->topic->id == topic_id) {
+            sub = cur;
+            break;
+        }
+    }
+
+    /* send PUBACK if needed (QoS > 0 or on invalid topic ID) */
+    uint8_t ret = (sub) ? MQTTSN_ACCEPTED : MQTTSN_REJ_INV_TOPIC_ID;
+    if ((sub == NULL) || (data[pos + 1] & MQTTSN_QOS_1)) {
+        uint8_t pkt[7] = { 7, MQTTSN_PUBACK, 0, 0, 0, 0, ret };
+        /* copy topic and message id */
+        memcpy(&pkt[2], &data[pos + 2], 4);
+        sock_udp_send(&con->sock, pkt, 7, &con->server_ep);
+    }
+
+    /* release the context and notify the user (on success) */
+    mutex_unlock(&con->lock);
+    if (sub) {
+        sub->cb(sub, ASYMCUTE_PUBLISHED,
+                &data[pos + 6], (len - (pos + 6)), sub->arg);
+    }
+}
+
+static void _on_puback(asymcute_con_t *con, const uint8_t *data, size_t len)
+{
+    mutex_lock(&con->lock);
+    asymcute_req_t *req = _req_preprocess(con, len, 7, data, 4);
+    if (req == NULL) {
+        mutex_unlock(&con->lock);
+        return;
+    }
+
+    unsigned ret = (data[6] == MQTTSN_ACCEPTED) ?
+                    ASYMCUTE_PUBLISHED : ASYMCUTE_REJECTED;
+    mutex_unlock(&req->lock);
+    mutex_unlock(&con->lock);
+    con->user_cb(req, ret);
+}
+
+static void _on_suback(asymcute_con_t *con, const uint8_t *data, size_t len)
+{
+    mutex_lock(&con->lock);
+    asymcute_req_t *req = _req_preprocess(con, len, 8, data, 5);
+    if (req == NULL) {
+        mutex_unlock(&con->lock);
+        return;
+    }
+
+    unsigned ret = ASYMCUTE_REJECTED;
+    if (data[7] == MQTTSN_ACCEPTED) {
+        /* parse and apply assigned topic id */
+        asymcute_sub_t *sub = (asymcute_sub_t *)req->arg;
+        sub->topic->id = byteorder_lebuftohs(&data[3]);
+        sub->topic->con = con;
+        /* insert subscription to connection context */
+        sub->next = con->subscriptions;
+        con->subscriptions = sub;
+        ret = ASYMCUTE_SUBSCRIBED;
+    }
+
+    /* notify the user */
+    mutex_unlock(&req->lock);
+    mutex_unlock(&con->lock);
+    con->user_cb(req, ret);
+}
+
+static void _on_unsuback(asymcute_con_t *con, const uint8_t *data, size_t len)
+{
+    mutex_lock(&con->lock);
+    asymcute_req_t *req = _req_preprocess(con, len, 4, data, 2);
+    if (req == NULL) {
+        mutex_unlock(&con->lock);
+        return;
+    }
+
+    /* remove subscription from list */
+    asymcute_sub_t *sub = (asymcute_sub_t *)req->arg;
+    if (con->subscriptions == sub) {
+        con->subscriptions = sub->next;
+    }
+    for (asymcute_sub_t *e = con->subscriptions; e->next; e = e->next) {
```

Comment:
Getting a segmentation fault when trying to unsubscribe from a (subscribed to) topic in the example application, since `e == NULL`, when `e->next` is checked.